File: internal/core/keys/keys.go
package keys

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"
	"time"
	"unicode"
)

const maxTimestampMillis int64 = 9999999999999999

type Builder struct {
	namespace string
}

func NewBuilder(namespace string) Builder {
	return Builder{namespace: normalizeNamespace(namespace)}
}

func (builder Builder) MessageBody(messageID string, createdAt time.Time) string {
	return builder.withNamespace(fmt.Sprintf("messages/%s/%s.json", TimePrefix(createdAt), messageID))
}

func (builder Builder) MessageLookup(messageID string) string {
	return builder.withNamespace(fmt.Sprintf("messages/by-id/%s.json", messageID))
}

func (builder Builder) AttachmentMetadata(attachmentID string, createdAt time.Time) string {
	return builder.withNamespace(fmt.Sprintf("attachments/metadata/%s/%s.json", TimePrefix(createdAt), attachmentID))
}

func (builder Builder) AttachmentLookup(attachmentID string) string {
	return builder.withNamespace(fmt.Sprintf("attachments/by-id/%s.json", attachmentID))
}

func (builder Builder) AttachmentObject(attachmentID string, createdAt time.Time, normalizedFileName string) string {
	return builder.withNamespace(fmt.Sprintf("attachments/objects/%s/%s/%s", TimePrefix(createdAt), attachmentID, NormalizeExternalID(normalizedFileName)))
}

func (builder Builder) MailboxReference(actorID string, direction string, createdAt time.Time, messageID string, referenceID string) string {
	return builder.withNamespace(fmt.Sprintf(
		"mailboxes/%s/%s/%s/%s_%s_%s_%s.json",
		NormalizeExternalID(actorID),
		NormalizeExternalID(direction),
		TimePrefix(createdAt),
		FeedSortKey(createdAt),
		CompactTimestamp(createdAt),
		messageID,
		referenceID,
	))
}

func (builder Builder) MailboxPrefix(actorID string, direction string, window time.Time) string {
	return builder.withNamespace(fmt.Sprintf(
		"mailboxes/%s/%s/%s/",
		NormalizeExternalID(actorID),
		NormalizeExternalID(direction),
		TimePrefix(window),
	))
}

func (builder Builder) ThreadMetadata(threadID string) string {
	return builder.withNamespace(fmt.Sprintf("threads/%s/metadata.json", threadID))
}

func (builder Builder) ThreadReference(threadID string, createdAt time.Time, messageID string, referenceID string) string {
	return builder.withNamespace(fmt.Sprintf(
		"threads/%s/messages/%s/%s_%s_%s_%s.json",
		threadID,
		TimePrefix(createdAt),
		ThreadSortKey(createdAt),
		CompactTimestamp(createdAt),
		messageID,
		referenceID,
	))
}

func (builder Builder) ThreadPrefix(threadID string, window time.Time) string {
	return builder.withNamespace(fmt.Sprintf("threads/%s/messages/%s/", threadID, TimePrefix(window)))
}

func (builder Builder) BroadcastBody(broadcastID string, createdAt time.Time) string {
	return builder.withNamespace(fmt.Sprintf("broadcast/messages/%s/%s.json", TimePrefix(createdAt), broadcastID))
}

func (builder Builder) BroadcastLookup(broadcastID string) string {
	return builder.withNamespace(fmt.Sprintf("broadcast/by-id/%s.json", broadcastID))
}

func (builder Builder) BroadcastAudience(audienceType string, audienceKey string, createdAt time.Time, broadcastID string) string {
	return builder.withNamespace(fmt.Sprintf(
		"broadcast/audiences/%s/%s/%s/%s_%s.json",
		NormalizeExternalID(audienceType),
		NormalizeExternalID(audienceKey),
		TimePrefix(createdAt),
		FeedSortKey(createdAt),
		broadcastID,
	))
}

func (builder Builder) BroadcastAudiencePrefix(audienceType string, audienceKey string, window time.Time) string {
	return builder.withNamespace(fmt.Sprintf(
		"broadcast/audiences/%s/%s/%s/",
		NormalizeExternalID(audienceType),
		NormalizeExternalID(audienceKey),
		TimePrefix(window),
	))
}

func (builder Builder) StateEvent(actorID string, targetKind string, targetID string, createdAt time.Time, stateID string) string {
	return builder.withNamespace(fmt.Sprintf(
		"states/%s/events/%s/%s/%s/%s_%s.json",
		NormalizeExternalID(actorID),
		NormalizeExternalID(targetKind),
		NormalizeExternalID(targetID),
		TimePrefix(createdAt),
		FeedSortKey(createdAt),
		stateID,
	))
}

func (builder Builder) StateCurrent(actorID string, targetKind string, targetID string) string {
	return builder.withNamespace(fmt.Sprintf(
		"states/%s/current/%s/%s.json",
		NormalizeExternalID(actorID),
		NormalizeExternalID(targetKind),
		NormalizeExternalID(targetID),
	))
}

func (builder Builder) OperationID(idempotencyKey string, callerID string) string {
	return builder.withNamespace(fmt.Sprintf(
		"operations/idempotency/%s/%s.json",
		NormalizeExternalID(callerID),
		NormalizeExternalID(idempotencyKey),
	))
}

func (builder Builder) OperationStarted(operationID string) string {
	return builder.withNamespace(fmt.Sprintf("operations/by-id/%s/started.json", operationID))
}

func (builder Builder) OperationStep(operationID string, stepID string) string {
	return builder.withNamespace(fmt.Sprintf("operations/by-id/%s/steps/%s.json", operationID, NormalizeExternalID(stepID)))
}

func (builder Builder) OperationCompleted(operationID string) string {
	return builder.withNamespace(fmt.Sprintf("operations/by-id/%s/completed.json", operationID))
}

func (builder Builder) withNamespace(key string) string {
	if builder.namespace == "" {
		return key
	}
	return builder.namespace + "/" + strings.TrimLeft(key, "/")
}

func TimePrefix(instant time.Time) string {
	utc := instant.UTC()
	return fmt.Sprintf("year=%04d/month=%02d/day=%02d/hour=%02d/minute=%02d",
		utc.Year(), int(utc.Month()), utc.Day(), utc.Hour(), utc.Minute())
}

func CompactTimestamp(instant time.Time) string {
	return instant.UTC().Format("20060102T150405.000000000Z")
}

func FeedSortKey(instant time.Time) string {
	millis := instant.UTC().UnixNano() / int64(time.Millisecond)
	return fmt.Sprintf("%016d", maxTimestampMillis-millis)
}

func ThreadSortKey(instant time.Time) string {
	millis := instant.UTC().UnixNano() / int64(time.Millisecond)
	return fmt.Sprintf("%016d", millis)
}

func NormalizeExternalID(raw string) string {
	trimmed := strings.TrimSpace(raw)
	if trimmed == "" {
		return "empty-" + shortHash(raw)
	}

	var builder strings.Builder
	previousDash := false
	for _, character := range strings.ToLower(trimmed) {
		valid := unicode.IsLetter(character) || unicode.IsDigit(character) || character == '-' || character == '_' || character == '.'
		if valid {
			builder.WriteRune(character)
			previousDash = false
			continue
		}
		if !previousDash {
			builder.WriteByte('-')
			previousDash = true
		}
	}

	normalized := strings.Trim(builder.String(), "-.")
	if normalized == "" {
		normalized = "id"
	}
	if len(normalized) > 96 {
		normalized = normalized[:72] + "-" + shortHash(trimmed)
	}
	return normalized
}

func normalizeNamespace(raw string) string {
	trimmed := strings.Trim(strings.TrimSpace(raw), "/")
	if trimmed == "" {
		return ""
	}
	return NormalizeExternalID(trimmed)
}

func shortHash(value string) string {
	sum := sha256.Sum256([]byte(value))
	return hex.EncodeToString(sum[:])[:16]
}

File: src/keys.rs
use chrono::{DateTime, Utc};
use sha2::{Digest, Sha256};

const MAX_TIMESTAMP_MILLIS: i64 = 9_999_999_999_999_999;
const MAX_NORMALIZED_SEGMENT_BYTES: usize = 96;
const HASH_SUFFIX_BYTES: usize = 16;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct KeyBuilder {
    namespace: String,
}

impl KeyBuilder {
    pub fn new(namespace: &str) -> Self {
        Self {
            namespace: normalize_namespace(namespace),
        }
    }

    pub fn message_body(&self, message_id: &str, created_at: DateTime<Utc>) -> String {
        self.with_namespace(&format!(
            "messages/{}/{}.json",
            time_prefix(created_at),
            message_id
        ))
    }

    pub fn message_lookup(&self, message_id: &str) -> String {
        self.with_namespace(&format!("messages/by-id/{}.json", message_id))
    }

    pub fn attachment_metadata(&self, attachment_id: &str, created_at: DateTime<Utc>) -> String {
        self.with_namespace(&format!(
            "attachments/metadata/{}/{}.json",
            time_prefix(created_at),
            attachment_id
        ))
    }

    pub fn attachment_lookup(&self, attachment_id: &str) -> String {
        self.with_namespace(&format!("attachments/by-id/{}.json", attachment_id))
    }

    pub fn attachment_object(
        &self,
        attachment_id: &str,
        created_at: DateTime<Utc>,
        normalized_file_name: &str,
    ) -> String {
        self.with_namespace(&format!(
            "attachments/objects/{}/{}/{}",
            time_prefix(created_at),
            attachment_id,
            normalize_file_name(normalized_file_name)
        ))
    }

    pub fn mailbox_reference(
        &self,
        actor_id: &str,
        direction: &str,
        created_at: DateTime<Utc>,
        message_id: &str,
        reference_id: &str,
    ) -> String {
        self.with_namespace(&format!(
            "mailboxes/{}/{}/{}/{}_{}_{}_{}.json",
            normalize_external_id(actor_id),
            normalize_external_id(direction),
            time_prefix(created_at),
            feed_sort_key(created_at),
            compact_timestamp(created_at),
            message_id,
            reference_id
        ))
    }

    pub fn mailbox_prefix(&self, actor_id: &str, direction: &str, window: DateTime<Utc>) -> String {
        self.with_namespace(&format!(
            "mailboxes/{}/{}/{}/",
            normalize_external_id(actor_id),
            normalize_external_id(direction),
            time_prefix(window)
        ))
    }

    pub fn thread_metadata(&self, thread_id: &str) -> String {
        self.with_namespace(&format!("threads/{}/metadata.json", thread_id))
    }

    pub fn thread_reference(
        &self,
        thread_id: &str,
        created_at: DateTime<Utc>,
        message_id: &str,
        reference_id: &str,
    ) -> String {
        self.with_namespace(&format!(
            "threads/{}/messages/{}/{}_{}_{}_{}.json",
            thread_id,
            time_prefix(created_at),
            thread_sort_key(created_at),
            compact_timestamp(created_at),
            message_id,
            reference_id
        ))
    }

    pub fn thread_prefix(&self, thread_id: &str, window: DateTime<Utc>) -> String {
        self.with_namespace(&format!(
            "threads/{}/messages/{}/",
            thread_id,
            time_prefix(window)
        ))
    }

    pub fn broadcast_body(&self, broadcast_id: &str, created_at: DateTime<Utc>) -> String {
        self.with_namespace(&format!(
            "broadcast/messages/{}/{}.json",
            time_prefix(created_at),
            broadcast_id
        ))
    }

    pub fn broadcast_lookup(&self, broadcast_id: &str) -> String {
        self.with_namespace(&format!("broadcast/by-id/{}.json", broadcast_id))
    }

    pub fn broadcast_audience(
        &self,
        audience_type: &str,
        audience_key: &str,
        created_at: DateTime<Utc>,
        broadcast_id: &str,
    ) -> String {
        self.with_namespace(&format!(
            "broadcast/audiences/{}/{}/{}/{}_{}.json",
            normalize_external_id(audience_type),
            normalize_external_id(audience_key),
            time_prefix(created_at),
            feed_sort_key(created_at),
            broadcast_id
        ))
    }

    pub fn broadcast_audience_prefix(
        &self,
        audience_type: &str,
        audience_key: &str,
        window: DateTime<Utc>,
    ) -> String {
        self.with_namespace(&format!(
            "broadcast/audiences/{}/{}/{}/",
            normalize_external_id(audience_type),
            normalize_external_id(audience_key),
            time_prefix(window)
        ))
    }

    pub fn state_event(
        &self,
        actor_id: &str,
        target_kind: &str,
        target_id: &str,
        created_at: DateTime<Utc>,
        state_id: &str,
    ) -> String {
        self.with_namespace(&format!(
            "states/{}/events/{}/{}/{}/{}_{}.json",
            normalize_external_id(actor_id),
            normalize_external_id(target_kind),
            normalize_external_id(target_id),
            time_prefix(created_at),
            feed_sort_key(created_at),
            state_id
        ))
    }

    pub fn state_current(&self, actor_id: &str, target_kind: &str, target_id: &str) -> String {
        self.with_namespace(&format!(
            "states/{}/current/{}/{}.json",
            normalize_external_id(actor_id),
            normalize_external_id(target_kind),
            normalize_external_id(target_id)
        ))
    }

    pub fn operation_id(&self, idempotency_key: &str, caller_id: &str) -> String {
        self.with_namespace(&format!(
            "operations/idempotency/{}/{}.json",
            normalize_external_id(caller_id),
            normalize_external_id(idempotency_key)
        ))
    }

    pub fn operation_started(&self, operation_id: &str) -> String {
        self.with_namespace(&format!("operations/by-id/{}/started.json", operation_id))
    }

    pub fn operation_step(&self, operation_id: &str, step_id: &str) -> String {
        self.with_namespace(&format!(
            "operations/by-id/{}/steps/{}.json",
            operation_id,
            normalize_external_id(step_id)
        ))
    }

    pub fn operation_completed(&self, operation_id: &str) -> String {
        self.with_namespace(&format!("operations/by-id/{}/completed.json", operation_id))
    }

    fn with_namespace(&self, key: &str) -> String {
        if self.namespace.is_empty() {
            key.to_string()
        } else {
            format!("{}/{}", self.namespace, key.trim_start_matches('/'))
        }
    }
}

pub fn time_prefix(instant: DateTime<Utc>) -> String {
    instant
        .with_timezone(&Utc)
        .format("year=%Y/month=%m/day=%d/hour=%H/minute=%M")
        .to_string()
}

pub fn compact_timestamp(instant: DateTime<Utc>) -> String {
    instant
        .with_timezone(&Utc)
        .format("%Y%m%dT%H%M%S.%fZ")
        .to_string()
}

pub fn feed_sort_key(instant: DateTime<Utc>) -> String {
    format!(
        "{:016}",
        MAX_TIMESTAMP_MILLIS - instant.with_timezone(&Utc).timestamp_millis()
    )
}

pub fn thread_sort_key(instant: DateTime<Utc>) -> String {
    format!("{:016}", instant.with_timezone(&Utc).timestamp_millis())
}

pub fn normalize_external_id(raw: &str) -> String {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return format!("empty-{}", short_hash(raw));
    }

    let normalized = sanitize_key_segment(trimmed, "id");
    let lossy = normalized != trimmed || normalized.len() > MAX_NORMALIZED_SEGMENT_BYTES;
    if lossy {
        with_hash_suffix(&normalized, trimmed)
    } else {
        normalized
    }
}

pub fn normalize_file_name(raw: &str) -> String {
    let normalized = sanitize_key_segment(raw.trim(), "file");
    if normalized.len() > MAX_NORMALIZED_SEGMENT_BYTES {
        with_hash_suffix(&normalized, raw.trim())
    } else {
        normalized
    }
}

fn normalize_namespace(raw: &str) -> String {
    let trimmed = raw.trim().trim_matches('/');
    if trimmed.is_empty() {
        String::new()
    } else {
        normalize_file_name(trimmed)
    }
}

fn sanitize_key_segment(raw: &str, fallback: &str) -> String {
    let mut normalized = String::new();
    let mut previous_dash = false;
    for character in raw.chars() {
        for lowered in character.to_lowercase() {
            let valid = lowered.is_alphanumeric() || matches!(lowered, '-' | '_' | '.');
            if valid {
                normalized.push(lowered);
                previous_dash = false;
            } else if !previous_dash {
                normalized.push('-');
                previous_dash = true;
            }
        }
    }

    let normalized = normalized.trim_matches(['-', '.']).to_string();
    if normalized.is_empty() {
        fallback.to_string()
    } else {
        normalized
    }
}

fn with_hash_suffix(normalized: &str, raw: &str) -> String {
    let hash = short_hash(raw);
    let max_prefix_bytes = MAX_NORMALIZED_SEGMENT_BYTES - HASH_SUFFIX_BYTES - 1;
    let prefix = truncate_to_boundary(normalized, max_prefix_bytes).trim_matches(['-', '.']);
    let prefix = if prefix.is_empty() { "id" } else { prefix };
    format!("{prefix}-{hash}")
}

fn truncate_to_boundary(value: &str, max_bytes: usize) -> &str {
    if value.len() <= max_bytes {
        return value;
    }
    let mut end = 0;
    for (index, character) in value.char_indices() {
        let next = index + character.len_utf8();
        if next > max_bytes {
            break;
        }
        end = next;
    }
    &value[..end]
}

fn short_hash(value: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(value.as_bytes());
    let digest = hasher.finalize();
    format!("{:x}", digest)[..16].to_string()
}

#[cfg(test)]
mod tests {
    use chrono::{Duration, TimeZone};

    use super::*;

    #[test]
    fn normalize_external_identifier() {
        assert_eq!(normalize_external_id("actor-a"), "actor-a");
        assert!(!normalize_external_id("a/b").contains('/'));
        assert_ne!(normalize_external_id("a/b"), normalize_external_id("a b"));
        assert_ne!(normalize_external_id("User"), normalize_external_id("user"));
        assert_eq!(normalize_file_name("Report Final.pdf"), "report-final.pdf");
    }

    #[test]
    fn time_prefix_and_sort_keys() {
        let older = Utc.with_ymd_and_hms(2026, 6, 1, 11, 22, 1).unwrap();
        let newer = older + Duration::minutes(1);

        assert_eq!(
            time_prefix(older),
            "year=2026/month=06/day=01/hour=11/minute=22"
        );
        assert!(feed_sort_key(newer) < feed_sort_key(older));
        assert!(thread_sort_key(older) < thread_sort_key(newer));
    }

    #[test]
    fn builder_applies_namespace() {
        let builder = KeyBuilder::new("dev/test");
        assert_eq!(
            builder.message_lookup("message-1"),
            "dev-test/messages/by-id/message-1.json"
        );
    }
}
